File: backend/app/pipeline.py
"""核心流水线：剧本 -> 分镜 -> 出图 -> 图生视频。

每一步都会即时把结果写回 storage，前端轮询 status 即可看到进度。
单镜失败不影响其他镜（标记 failed 并记录 error）。
"""
from __future__ import annotations

from pathlib import Path

from .config import OUTPUTS_DIR, DATA_DIR, get_settings
from .models import Project, Shot, ShotStatus
from .providers import registry
from . import storage


def _rel(path: Path) -> str:
    """转成相对 data/ 的路径，便于前端通过 /media/ 访问。"""
    return str(path.relative_to(DATA_DIR))


def analyze_script(project: Project) -> Project:
    """用 LLM（或 Mock）把剧本拆成分镜，写回 project.shots。"""
    settings = get_settings()
    llm = registry.get_llm(settings)
    drafts = llm.analyze_script(project.script, project.characters)

    name_to_id = {c.name: c.id for c in project.characters}
    shots: list[Shot] = []
    for i, d in enumerate(drafts):
        char_ids = [name_to_id[n] for n in d.character_names if n in name_to_id]
        shots.append(
            Shot(
                index=i,
                summary=d.summary,
                image_prompt=d.image_prompt,
                video_prompt=d.video_prompt,
                character_ids=char_ids,
            )
        )
    project.shots = shots
    return storage.save_project(project)


def _ref_paths(project: Project, shot: Shot) -> list[Path]:
    paths: list[Path] = []
    for c in project.characters:
        if c.id in shot.character_ids and c.ref_image:
            p = DATA_DIR / c.ref_image
            if p.exists():
                paths.append(p)
    return paths


def generate_image(project_id: str, shot_id: str) -> Shot:
    settings = get_settings()
    provider = registry.get_image(settings)
    project = storage.get_project(project_id)
    if project is None:
        raise ValueError("project not found")
    shot = _find_shot(project, shot_id)

    shot.status = ShotStatus.image_running
    shot.error = None
    storage.save_project(project)
    try:
        data = provider.generate(shot.image_prompt, _ref_paths(project, shot))
        out = OUTPUTS_DIR / f"{shot.id}.png"
        out.write_bytes(data)
        shot.image_path = _rel(out)
        shot.status = ShotStatus.image_done
    except Exception as exc:  # noqa: BLE001 - 单镜失败需记录而非中断
        shot.status = ShotStatus.failed
        shot.error = f"出图失败: {exc}"
    storage.save_project(project)
    return shot


def generate_video(project_id: str, shot_id: str) -> Shot:
    settings = get_settings()
    provider = registry.get_video(settings)
    project = storage.get_project(project_id)
    if project is None:
        raise ValueError("project not found")
    shot = _find_shot(project, shot_id)

    if not shot.image_path:
        shot.status = ShotStatus.failed
        shot.error = "请先出图再生成视频"
        storage.save_project(project)
        return shot

    shot.status = ShotStatus.video_running
    shot.error = None
    storage.save_project(project)
    try:
        image_path = DATA_DIR / shot.image_path
        data = provider.image_to_video(image_path, shot.video_prompt, settings.video_duration)
        out = OUTPUTS_DIR / f"{shot.id}.mp4"
        out.write_bytes(data)
        shot.video_path = _rel(out)
        shot.status = ShotStatus.done
    except Exception as exc:  # noqa: BLE001
        shot.status = ShotStatus.failed
        shot.error = f"图生视频失败: {exc}"
    storage.save_project(project)
    return shot


def generate_all(project_id: str) -> None:
    """一键生成：对每个分镜依次出图 + 图生视频。供后台任务调用。"""
    project = storage.get_project(project_id)
    if project is None:
        return
    for shot in project.shots:
        s = generate_image(project_id, shot.id)
        if s.status == ShotStatus.failed:
            continue
        generate_video(project_id, shot.id)


def _find_shot(project: Project, shot_id: str) -> Shot:
    for s in project.shots:
        if s.id == shot_id:
            return s
    raise ValueError("shot not found")

File: backend-go/internal/pipeline/pipeline.go
// Package pipeline 生成流水线 worker：剧本分析 / 批量出图 / 图生视频。
// 消费队列消息，调用 provider，结果写回 storage，并通过 ws.Hub 推送进度。
package pipeline

import (
	"context"
	"fmt"
	"log/slog"
	"os"
	"path/filepath"
	"time"

	"github.com/zuozuo0320/ai-film-studio/backend-go/internal/models"
	"github.com/zuozuo0320/ai-film-studio/backend-go/internal/provider"
	"github.com/zuozuo0320/ai-film-studio/backend-go/internal/queue"
	"github.com/zuozuo0320/ai-film-studio/backend-go/internal/storage"
	"github.com/zuozuo0320/ai-film-studio/backend-go/internal/ws"
)

type Pipeline struct {
	store    storage.Store
	hub      *ws.Hub
	llm      provider.LLM
	image    provider.Image
	video    provider.Video
	dataDir  string
	videoSec int
}

func New(store storage.Store, hub *ws.Hub, llm provider.LLM, img provider.Image, vid provider.Video, dataDir string, videoSec int) *Pipeline {
	return &Pipeline{store: store, hub: hub, llm: llm, image: img, video: vid, dataDir: dataDir, videoSec: videoSec}
}

// Register 把三类 worker 挂到队列上。
func (p *Pipeline) Register(q queue.Queue) {
	q.Subscribe(queue.TopicScriptAnalyze, p.handleAnalyze)
	q.Subscribe(queue.TopicImageBatch, p.handleImageBatch)
	q.Subscribe(queue.TopicVideoGenerate, p.handleVideo)
}

// ---------- 剧本分析 ----------

func (p *Pipeline) handleAnalyze(ctx context.Context, msg queue.Message) {
	proj, err := p.store.GetProject(msg.ProjectID)
	if err != nil {
		slog.Error("analyze: project not found", "project_id", msg.ProjectID)
		return
	}
	drafts, err := p.llm.AnalyzeScript(ctx, proj.Script, proj.Characters)
	if err != nil {
		slog.Error("analyze failed", "err", err)
		return
	}
	nameToID := map[string]string{}
	for _, c := range proj.Characters {
		nameToID[c.Name] = c.ID
	}
	shots := make([]models.Shot, 0, len(drafts))
	for i, d := range drafts {
		s := models.NewShot(i)
		s.Summary, s.ImagePrompt, s.Keywords, s.VideoPrompt = d.Summary, d.ImagePrompt, d.Keywords, d.VideoPrompt
		for _, n := range d.CharacterNames {
			if id, ok := nameToID[n]; ok {
				s.CharacterIDs = append(s.CharacterIDs, id)
			}
		}
		shots = append(shots, s)
	}
	proj.Shots = shots
	if err := p.store.UpdateProject(proj); err != nil {
		slog.Error("analyze: save failed", "err", err)
		return
	}
	p.hub.Publish(ws.Event{Type: "project_analyzed", ProjectID: proj.ID, Status: "done"})
}

// ---------- 批量出图 ----------

func (p *Pipeline) handleImageBatch(ctx context.Context, msg queue.Message) {
	batch, err := p.store.GetBatch(msg.BatchID)
	if err != nil {
		slog.Error("image batch not found", "batch_id", msg.BatchID)
		return
	}
	batch.Status = models.BatchRunning
	_ = p.store.UpdateBatch(batch)
	p.publishBatch(batch)

	failed := 0
	for _, shotID := range batch.ShotIDs {
		if err := p.generateImage(ctx, batch.ProjectID, shotID); err != nil {
			failed++
		}
	}
	if len(batch.ShotIDs) > 0 && failed == len(batch.ShotIDs) {
		batch.Status = models.BatchFailed
	} else {
		batch.Status = models.BatchDone
	}
	t := time.Now()
	batch.FinishedAt = &t
	_ = p.store.UpdateBatch(batch)
	p.publishBatch(batch)
}

func (p *Pipeline) generateImage(ctx context.Context, projectID, shotID string) error {
	proj, shot, err := p.findShot(projectID, shotID)
	if err != nil {
		return err
	}
	p.setShotStatus(proj, shot, models.ShotImageRunning, "")

	var refs []string
	for _, c := range proj.Characters {
		for _, id := range shot.CharacterIDs {
			if c.ID == id {
				refs = append(refs, c.RefImages...)
			}
		}
	}
	data, err := p.image.Generate(ctx, shot.ImagePrompt, refs)
	if err != nil {
		p.setShotStatus(proj, shot, models.ShotFailed, fmt.Sprintf("出图失败: %v", err))
		return err
	}
	rel := filepath.Join("outputs", shot.ID+".png")
	if err := p.writeFile(rel, data); err != nil {
		p.setShotStatus(proj, shot, models.ShotFailed, fmt.Sprintf("写文件失败: %v", err))
		return err
	}
	shot.ImagePath = rel
	p.setShotStatus(proj, shot, models.ShotImageReview, "")
	return nil
}

// ---------- 图生视频 ----------

func (p *Pipeline) handleVideo(ctx context.Context, msg queue.Message) {
	proj, shot, err := p.findShot(msg.ProjectID, msg.ShotID)
	if err != nil {
		slog.Error("video: shot not found", "shot_id", msg.ShotID)
		return
	}
	if shot.ImagePath == "" {
		p.setShotStatus(proj, shot, models.ShotFailed, "请先出图再生成视频")
		return
	}
	p.setShotStatus(proj, shot, models.ShotVideoRunning, "")

	data, err := p.video.ImageToVideo(ctx, filepath.Join(p.dataDir, shot.ImagePath), shot.VideoPrompt, p.videoSec)
	if err != nil {
		p.setShotStatus(proj, shot, models.ShotFailed, fmt.Sprintf("图生视频失败: %v", err))
		return
	}
	rel := filepath.Join("outputs", shot.ID+".mp4")
	if err := p.writeFile(rel, data); err != nil {
		p.setShotStatus(proj, shot, models.ShotFailed, fmt.Sprintf("写文件失败: %v", err))
		return
	}
	shot.VideoPath = rel
	p.setShotStatus(proj, shot, models.ShotDone, "")
}

// ---------- helpers ----------

func (p *Pipeline) findShot(projectID, shotID string) (*models.Project, *models.Shot, error) {
	proj, err := p.store.GetProject(projectID)
	if err != nil {
		return nil, nil, err
	}
	for i := range proj.Shots {
		if proj.Shots[i].ID == shotID {
			return proj, &proj.Shots[i], nil
		}
	}
	return nil, nil, storage.ErrNotFound
}

// setShotStatus 更新分镜状态并落库 + 推送事件。
// 注意 shot 是 proj.Shots 内的指针，UpdateProject 持久化整个项目。
func (p *Pipeline) setShotStatus(proj *models.Project, shot *models.Shot, st models.ShotStatus, errMsg string) {
	shot.Status = st
	shot.Error = errMsg
	if err := p.store.UpdateProject(proj); err != nil {
		slog.Error("save shot status failed", "err", err)
	}
	p.hub.Publish(ws.Event{Type: "shot_update", ProjectID: proj.ID, ShotID: shot.ID, Status: string(st), Error: errMsg})
}

func (p *Pipeline) publishBatch(b *models.GenerationBatch) {
	p.hub.Publish(ws.Event{Type: "batch_update", ProjectID: b.ProjectID, BatchID: b.ID, Status: string(b.Status)})
}

func (p *Pipeline) writeFile(rel string, data []byte) error {
	abs := filepath.Join(p.dataDir, rel)
	if err := os.MkdirAll(filepath.Dir(abs), 0o755); err != nil {
		return err
	}
	return os.WriteFile(abs, data, 0o644)
}
